File: app/src/main/java/com/sofascoreacademy/sofascoreacademy/MyDiffUtilCallbackTeamsPlayers.java
package com.sofascoreacademy.sofascoreacademy;

import androidx.recyclerview.widget.DiffUtil;

import com.sofascoreacademy.model.Team;
import com.sofascoreacademy.sofascoreacademy.model.Player;

import java.util.List;

public class MyDiffUtilCallbackTeamsPlayers extends DiffUtil.Callback {

    List<Object> oldlist;
    List<Object> newlist;

    public MyDiffUtilCallbackTeamsPlayers(List<Object> oldlist, List<Object> newlist) {
        this.oldlist = oldlist;
        this.newlist = newlist;
    }

    @Override
    public int getOldListSize() {
        return oldlist.size();
    }

    @Override
    public int getNewListSize() {
        return newlist.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Object objectOld = oldlist.get(oldItemPosition);
        Object objectNew = oldlist.get(newItemPosition);

        if (objectOld instanceof Team && objectNew instanceof Team) {
            Team teamOld = (Team) objectOld;
            Team teamNew = (Team) objectNew;

            return teamOld.getId() == teamNew.getId();
        } else if (objectOld instanceof Player && objectNew instanceof Player) {
            Player playerOld = (Player) objectOld;
            Player playerNew = (Player) objectNew;

            return playerOld.getId() == playerNew.getId();
        } else {
            return false;
        }
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return false;
    }
}

File: app/src/main/java/com/sofascoreacademy/sofascoreacademy/MyDiffUtilCallbackTeamsPlayers.kt
package com.sofascoreacademy.sofascoreacademy

import androidx.recyclerview.widget.DiffUtil
import com.sofascoreacademy.model.Team
import com.sofascoreacademy.sofascoreacademy.model.Player

class MyDiffUtilCallbackTeamsPlayers(var oldlist: List<Any>, var newlist: List<Any>) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldlist.size
    }

    override fun getNewListSize(): Int {
        return newlist.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val objectOld = oldlist[oldItemPosition]
        val objectNew = oldlist[newItemPosition]
        return if (objectOld is Team && objectNew is Team) {
            objectOld.id == objectNew.id
        } else if (objectOld is Player && objectNew is Player) {
            objectOld.id == objectNew.id
        } else {
            false
        }
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return false
    }

}